{"url":"https://api.github.com/repos/elastic/elasticsearch/issues/17482","repository_url":"https://api.github.com/repos/elastic/elasticsearch","labels_url":"https://api.github.com/repos/elastic/elasticsearch/issues/17482/labels{/name}","comments_url":"https://api.github.com/repos/elastic/elasticsearch/issues/17482/comments","events_url":"https://api.github.com/repos/elastic/elasticsearch/issues/17482/events","html_url":"https://github.com/elastic/elasticsearch/issues/17482","id":145236107,"node_id":"MDU6SXNzdWUxNDUyMzYxMDc=","number":17482,"title":"BulkProcessor exception in 2.3.0 - IllegalIndexShardStateException","user":{"login":"reflection","id":652093,"node_id":"MDQ6VXNlcjY1MjA5Mw==","avatar_url":"https://avatars2.githubusercontent.com/u/652093?v=4","gravatar_id":"","url":"https://api.github.com/users/reflection","html_url":"https://github.com/reflection","followers_url":"https://api.github.com/users/reflection/followers","following_url":"https://api.github.com/users/reflection/following{/other_user}","gists_url":"https://api.github.com/users/reflection/gists{/gist_id}","starred_url":"https://api.github.com/users/reflection/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/reflection/subscriptions","organizations_url":"https://api.github.com/users/reflection/orgs","repos_url":"https://api.github.com/users/reflection/repos","events_url":"https://api.github.com/users/reflection/events{/privacy}","received_events_url":"https://api.github.com/users/reflection/received_events","type":"User","site_admin":false},"labels":[{"id":145572580,"node_id":"MDU6TGFiZWwxNDU1NzI1ODA=","url":"https://api.github.com/repos/elastic/elasticsearch/labels/:Distributed/CRUD","name":":Distributed/CRUD","color":"0e8a16","default":false,"description":"A catch all label for issues around indexing, updating and getting a doc by id. Not search."}],"state":"closed","locked":false,"assignee":null,"assignees":[],"milestone":null,"comments":2,"created_at":"2016-04-01T17:24:06Z","updated_at":"2018-02-13T19:39:25Z","closed_at":"2016-04-04T23:28:56Z","author_association":"NONE","active_lock_reason":null,"body":"**Elasticsearch version**: 2.3.0 (official es docker image)\n\n**JVM version**: OpenJDK Runtime Environment (build 1.8.0_72-internal-b15)\n\n**OS version**: Debian 8\n\n**Description of the problem including expected versus actual behavior**: Starting from es 2.2 and still in 2.3, we get the following exception only while using the Java BulkProcessor:\n\n```\n[356019-20160321][[356019-20160321][0]] IllegalIndexShardStateException[CurrentState[POST_RECOVERY] operation only allowed when started/recovering, origin [PRIMARY]]\n    at org.elasticsearch.index.shard.IndexShard.ensureWriteAllowed(IndexShard.java:1068)\n    at org.elasticsearch.index.shard.IndexShard.create(IndexShard.java:538)\n    at org.elasticsearch.index.engine.Engine$Create.execute(Engine.java:810)\n    at org.elasticsearch.action.index.TransportIndexAction.executeIndexRequestOnPrimary(TransportIndexAction.java:237)\n    at org.elasticsearch.action.bulk.TransportShardBulkAction.shardIndexOperation(TransportShardBulkAction.java:326)\n    at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:119)\n    at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:68)\n    at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.doRun(TransportReplicationAction.java:639)\n    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)\n    at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:279)\n    at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:271)\n    at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:75)\n    at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)\n    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)\n    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n    at java.lang.Thread.run(Thread.java:745)\n```\n\nDid we lose the records that would have been indexed, or does BulkProcessor wait for the index to get in a state where it can write and retry?  If the latter, are these exceptions \"harmless\" as stated in #15478 (there were multiple exceptions there so not sure if this was considered the harmless one).\n\n**Steps to reproduce**:\n1. Consume from Kafka topic with three consumer threads that share one ES BulkProcessor object with all default values.\n","closed_by":{"login":"reflection","id":652093,"node_id":"MDQ6VXNlcjY1MjA5Mw==","avatar_url":"https://avatars2.githubusercontent.com/u/652093?v=4","gravatar_id":"","url":"https://api.github.com/users/reflection","html_url":"https://github.com/reflection","followers_url":"https://api.github.com/users/reflection/followers","following_url":"https://api.github.com/users/reflection/following{/other_user}","gists_url":"https://api.github.com/users/reflection/gists{/gist_id}","starred_url":"https://api.github.com/users/reflection/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/reflection/subscriptions","organizations_url":"https://api.github.com/users/reflection/orgs","repos_url":"https://api.github.com/users/reflection/repos","events_url":"https://api.github.com/users/reflection/events{/privacy}","received_events_url":"https://api.github.com/users/reflection/received_events","type":"User","site_admin":false},"performed_via_github_app":null}